numpy/_core/src/multiarray/array_converter.c

```diff
@@ -0,0 +1,558 @@
+#include "abstractdtypes.h"
+#include "common_dtype.h"
+#include "convert_datatype.h"
+#include "descriptor.h"
+#include "multiarraymodule.h"
+#include "ctors.h"
+
+#include "npy_config.h"
+
+#include "npy_pycompat.h"
+#include "array_assign.h"
+
+#include "common.h"
+#include "get_attr_string.h"
+
+
+
+static PyObject *
+array_converter_new(
+        PyTypeObject *cls, PyObject *args, PyObject *kwds)
+{
+    if (kwds != NULL && PyDict_GET_SIZE(kwds) != 0) {
+        PyErr_SetString(PyExc_TypeError,
+                "Array creation helper doesn't support keywords.");
+        return NULL;
+    }
+
+    Py_ssize_t narrs_ssize_t = (args == NULL) ? 0 : PyTuple_GET_SIZE(args);
+    int narrs = (int)narrs_ssize_t;
+    /* Limit to NPY_MAXARGS for now. */
+    if (narrs_ssize_t > NPY_MAXARGS) {
+        PyErr_SetString(PyExc_RuntimeError,
+            "too many arrays.");
+        return NULL;
+    }
+
+    PyArrayArrayConverterObject *self = PyObject_NewVar(
+            PyArrayArrayConverterObject, cls, narrs);
+    if (self == NULL) {
+        return NULL;
+    }
+    PyObject_InitVar((PyVarObject *)self, &PyArrayArrayConverter_Type, narrs);
+
+    self->narrs = 0;
+    self->flags = 0;
+    self->wrap = NULL;
+    self->wrap_type = NULL;
+
+    if (narrs == 0) {
+        return (PyObject *)self;
+    }
+    self->flags = (NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
+
+    creation_item *item = self->items;
+    /* increase self->narrs in loop for cleanup */
+    for (int i = 0; i < narrs; i++, item++) {
+        item->object = PyTuple_GET_ITEM(args, i);
+
+        /* Fast path if input is an array (maybe FromAny should be faster): */
+        if (PyArray_Check(item->object)) {
+            Py_INCREF(item->object);
+            item->array = (PyArrayObject *)item->object;
+            item->scalar_input = 0;
+        }
+        else {
+            item->array = (PyArrayObject *)PyArray_FromAny_int(
+                    item->object, NULL, NULL, 0, 0, 0, NULL,
+                    &item->scalar_input);
+            if (item->array == NULL) {
+                goto fail;
+            }
+        }
+        if (!item->scalar_input) {
+            self->flags &= ~(NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
+        }
+        /* At this point, assume cleanup should happen for this item */
+        self->narrs++;
+        Py_INCREF(item->object);
+        item->DType = NPY_DTYPE(PyArray_DESCR(item->array));
+        Py_INCREF(item->DType);
+
+        if (npy_mark_tmp_array_if_pyscalar(
+                item->object, item->array, &item->DType)) {
+            /* just clear the flag again */
+            ((PyArrayObject_fields *)(item->array))->flags &= (
+                    ~NPY_ARRAY_WAS_PYTHON_LITERAL);
+            item->descr = NULL;
+        }
+        else {
+            self->flags &= ~NPY_CH_ALL_PYSCALARS;
+
+            item->descr = PyArray_DESCR(item->array);
+            Py_INCREF(item->descr);
+        }
+    }
+
+    Py_INCREF(self);
```

Comment:
Are you sure you need the `INCREF` here? `PyObject_NewVar` returns a new reference already (which you `DECREF` in the fail path).